### DISCUSSION TITLE
svgo配置不生效

### DISCUSSION THREAD

**Discussion by Jesuiswkai** [umijs/umi#9342] [Q&A]
4.x版本把svg默认data改成了base64，我改了svg的配置也没有生效。

![umirc](https://user-images.githubusercontent.com/68602043/190545239-71cc4b8e-caac-4eb0-be71-90bc7eda1189.png)
![base64](https://user-images.githubusercontent.com/68602043/190545259-2581d204-6658-4eac-b6cf-891b897a30b1.png)

**Comment by fz6m** [comment]
给一个复现看看。

**Comment by Jeff-Tian** [comment]
Hi 我也碰到怀疑是“同样”的问题，我在 umi 中集成 ckeditor5，toolbar 显示不出来，原因是 svg 被强行转化成了base64，并且没有 viewBox 属性导致的。尝试通过 svgo 配置 unenc，**无效。请大佬们帮忙看看什么原因？**

<img width="1524" alt="image" src="https://github.com/umijs/umi/assets/3367820/f663fd38-26e3-400f-883d-6ad0bdbb4da2">

复现仓库：https://github.com/JeffTrain/umi-ckeditor5

`yarn && yarn dev ` 即可复现。

## 正常结果
（不用 umi4，用 umi3 或者 cra，都可以正常显示 toolbar）

![v2-383c371be1a358119fa12f247eadab6a](https://github.com/umijs/umi/assets/3367820/39b22214-e877-40d3-a505-a46798bc1654)

**Reply by fz6m** [reply]
项目里对 webpack 的改写太多了，hack 了太多逻辑，难免会有匹配出错的问题或者没考虑到的情况，需要花费很多时间，没法看，如有问题，需要一个最小的复现。

**Comment by Jeff-Tian** [comment]
感谢大佬回复！我重新复现了一下，没有任何额外 hack，代码提交到了 https://github.com/orgs/JeffTrain/packages?repo_name=umijs-ckeditor5。

- 先创建 umi 项目： pnpm dlx create-umi@latest
```
pnpm dlx create-umi@latest
 WARN  GET https://r.cnpmjs.org/glob-parent error (ETIMEDOUT). Will retry in 10 seconds. 2 retries left.
 WARN  GET https://r.cnpmjs.org/on-exit-leak-free error (ETIMEDOUT). Will retry in 10 seconds. 2 retries left.
Packages: +39
+++++++++++++++++++++++++++++++++++++++
Progress: resolved 40, reused 38, downloaded 1, added 39, done
┌   create-umi
│
○  Pick Umi App Template
│  Ant Design Pro
│
○  Pick Npm Client
│  pnpm
│
○  Pick Npm Registry
│  taobao
│
└  You're all set!
```
- 然后在 package.json 里添加 ckeditor 依赖，pnpm install && pnpm dev
- 接着，在打开的浏览器里打开开发者工具，可以看到 CKEditor 尝试初始化 UI，并报错。是尝试获取 viewbox 时，得到了一个 base64 并且没有 viewbox 属性的 content：
![微信截图_20231024090258](https://github.com/umijs/umi/assets/3367820/7e637128-2bd3-40f2-a24e-d4d5510ddcbb)

**Reply by fz6m** [reply]
```bash
  pnpm add -D postcss-loader style-loader css-loader
```

```ts
// .umirc.ts

  chainWebpack(config) {
    const { styles } = require('@ckeditor/ckeditor5-dev-utils');
    const svgReg = /ckeditor5-[^/\\]+[/\\]theme[/\\]icons[/\\][^/\\]+\.svg$/;
    const cssReg = /ckeditor5-[^/\\]+[/\\]theme[/\\].+\.css$/;
    config.module.rule('cke5-svg').test(svgReg).type(
      // raw-loader
      'asset/source',
    );
    // svg exclude
    ['svg', 'svgr'].forEach((rule) => {
      config.module.rule(rule).exclude.add(svgReg);
    });
    // css rule
    config.module
      .rule('cke5-css')
      .test(cssReg)
      .use('style-loader')
      .loader(require.resolve('style-loader'))
      .end()
      .use('css-loader')
      .loader(require.resolve('css-loader'))
      .end()
      .use('postcss-loader')
      .loader(require.resolve('postcss-loader'))
      .options({
        postcssOptions: styles.getPostCssConfig({
          themeImporter: {
            themePath: require.resolve('@ckeditor/ckeditor5-theme-lark'),
          },
          minify: true,
        }),
      });
    // css exclude
    config.module.rule('css').exclude.add(cssReg);
    // assets exclude
    config.module
      .rule('asset')
      .oneOf('fallback')
      .exclude.add(svgReg)
      .add(cssReg);
  },
```

**Reply by Jeff-Tian** [reply]
umi3 没有这个问题，我这里做了一个 demo： https://github.com/Jeff-Tian/umi-ckeditor5 。

<img width="1452" alt="image" src="https://github.com/umijs/umi/assets/3367820/c47153de-2494-492b-a00b-c486b5f743d3">

<img width="778" alt="image" src="https://github.com/umijs/umi/assets/3367820/c9a6ec2d-42c5-4b0d-b46c-069708e5f38c">

[a comment by an automated account is not shown]

**Comment by Jeff-Tian** [comment]
> ```shell
>   pnpm add -D postcss-loader style-loader css-loader
> ```
> 
> ```ts
> // .umirc.ts
> 
>   chainWebpack(config) {
>     const { styles } = require('@ckeditor/ckeditor5-dev-utils');
>     const svgReg = /ckeditor5-[^/\\]+[/\\]theme[/\\]icons[/\\][^/\\]+\.svg$/;
>     const cssReg = /ckeditor5-[^/\\]+[/\\]theme[/\\].+\.css$/;
>     config.module.rule('cke5-svg').test(svgReg).type(
>       // raw-loader
>       'asset/source',
>     );
>     // svg exclude
>     ['svg', 'svgr'].forEach((rule) => {
>       config.module.rule(rule).exclude.add(svgReg);
>     });
>     // css rule
>     config.module
>       .rule('cke5-css')
>       .test(cssReg)
>       .use('style-loader')
>       .loader(require.resolve('style-loader'))
>       .end()
>       .use('css-loader')
>       .loader(require.resolve('css-loader'))
>       .end()
>       .use('postcss-loader')
>       .loader(require.resolve('postcss-loader'))
>       .options({
>         postcssOptions: styles.getPostCssConfig({
>           themeImporter: {
>             themePath: require.resolve('@ckeditor/ckeditor5-theme-lark'),
>           },
>           minify: true,
>         }),
>       });
>     // css exclude
>     config.module.rule('css').exclude.add(cssReg);
>     // assets exclude
>     config.module
>       .rule('asset')
>       .oneOf('fallback')
>       .exclude.add(svgReg)
>       .add(cssReg);
>   },
> ```

感谢大佬的回复，应用了这个改动，svg 依然是被优化成了 base64。提交见： https://github.com/JeffTrain/umijs-ckeditor5/commit/956a61e94246463b6a828731e9eb6c66622529be

<img width="1495" alt="image" src="https://github.com/umijs/umi/assets/3367820/da8ac668-4caf-4531-93af-424032dbc32f">

**Reply by fz6m** [reply]
你说的 svg 是哪来的，我在你这个项目里配置了上面我贴的配置，之后可以正常启动项目运行：

<img src='https://github.com/umijs/umi/assets/59400654/2f52fa33-3576-4bea-b47e-3ea38374cb38' width='50%' />

**Reply by Jeff-Tian** [reply]
感谢大佬！我将本地的 pnpm 从 6.11.0 升级到→ 8.9.2 (`pnpm add -g pnpm` ) ，然后重新 `pnpm i` 后，真的好了！ 看来是本地有些依赖的版本的问题（https://github.com/JeffTrain/umijs-ckeditor5/commit/db3e85e13b219264cae1bcdadd6344e05dcb1a24 ）？

不过  toolbar 上的 svg 现在正常显示了，非常感谢！

**Reply by LoveRecall** [reply]
@fz6m 按照你的配置，编辑器出来了，可是默认样式没有了，我看了是完全没有加载出来，而不是被其他css覆盖了
![image](https://github.com/umijs/umi/assets/17865339/f09524ca-5b90-4eb7-a052-5481d4952d0a)

**Reply by fz6m** [reply]
如有问题，开一个新的 issue ，需要给一个最小复现帮助排查问题。
